frame/staking/src/tests.rs

```diff
@@ -3827,88 +3827,103 @@ fn on_finalize_weight_is_nonzero() {
 
 // end-to-end nodes of the voter bags operation.
 mod voter_bags {
+	use super::Origin;
+	use crate::{mock::*, ValidatorPrefs};
+	use frame_support::{assert_ok, traits::Currency};
 
 	#[test]
-	fn rebag_works() {
-		todo!()
-	}
-}
-/*
-// TODO: this needs some love, retire it in favour of the one above. Use the mock data, don't make
-// it complicated with data setup, use the simplest data possible, instead check multiple
-// edge-cases.
-#[test]
-fn test_rebag() {
-	use crate::{
-		testing_utils::create_stash_controller,
-		voter_bags::{Bag, Node},
-	};
-	use frame_system::RawOrigin;
-
-	/// Make a validator and return its stash
-	fn make_validator(n: u32, balance_factor: u32) -> Result<AccountIdOf<Test>, &'static str> {
-		let (stash, controller) = create_stash_controller::<Test>(n, balance_factor, Default::default()).unwrap();
-
-		// Bond the full value of the stash
-		//
-		// By default, `create_stash_controller` only bonds 10% of the stash. However, we're going
-		// to want to edit one account's bonded value to match another's, so it's simpler if 100% of
-		// the balance is bonded.
-		let balance = <Test as Config>::Currency::free_balance(&stash);
-		Staking::bond_extra(RawOrigin::Signed(stash.clone()).into(), balance).unwrap();
-		Staking::validate(
-			RawOrigin::Signed(controller.clone()).into(),
-			ValidatorPrefs::default(),
-		).unwrap();
-
-		Ok(stash)
+	fn insert_and_remove_works() {
+		// we test insert/remove indirectly via `validate`, `nominate`, and chill
+		ExtBuilder::default().build_and_execute(|| {
+			// given
+			assert_eq!(get_bags(), vec![(10, vec![31]), (1000, vec![11, 21, 101])]);
+
+			// `bond`
+			bond(42, 43, 2_000);
+			// does not insert the voter
+			assert_eq!(get_bags(), vec![(10, vec![31]), (1000, vec![11, 21, 101])]);
+
+			// `validate`
+			assert_ok!(Staking::validate(Origin::signed(43).into(), ValidatorPrefs::default()));
+			// moves the voter into a bag
+			assert_eq!(
+				get_bags(),
+				vec![(10, vec![31]), (1000, vec![11, 21, 101]), (2000, vec![42])]
+			);
+
+			// `nominate`-ing, but not changing active stake (which implicitly calls remove)
+			assert_ok!(Staking::nominate(Origin::signed(43), vec![11]));
+			// does not change the voters position
+			assert_eq!(
+				get_bags(),
+				vec![(10, vec![31]), (1000, vec![11, 21, 101]), (2000, vec![42])]
+			);
+
+			// `chill`
+			assert_ok!(Staking::chill(Origin::signed(43)));
+			// removes the voter
+			assert_eq!(get_bags(), vec![(10, vec![31]), (1000, vec![11, 21, 101]), (2000, vec![])]);
```

Comment:
seems like we have a stale storage entry here, the 2000 bag which is empty. not a huge deal, but ideally once we remove the last item from the bag, it should be totally cleared.

Reply:
fixed it in [`e7fdad2` (#9422)](https://github.com/paritytech/substrate/pull/9422/commits/e7fdad2aa55c5e71ee209a11baf8b3014248f260)

Reply:
remaining tests fixed in 724c17ef1bf4005a033c04ee1e5d8e4d79240714